knative/eventing-operator
# Reconciliation for default broker selection

# Issue to be fixed

Fixes #175

## Proposed Changes

* New `DefaultBrokerClass` in KnativeEventing CR
* Make the given broker default broker

Notes:
* Overwrites the `default-br-config` field of the `config-br-defaults` configmap based on the value given on CR.
* If the configmap already exists, it will still be overridden. This behavior is like that for everything, not specific to this configmap. For example, if a deployment/configmap/whatever that's defined in the manifest exists, it will be overwritten by Manifestival. I didn't do anything special for this configmap.
* I didn't want to introduce dependency to `knative/eventing`, but in the end it was super bad to duplicate the parsing logic of the configmap's data (YAML in string) and I decided to reuse some functions from `knative/eventing`. The functions used are specific to eventing only and cannot be really moved to `knative/pkg`.


Verification instructions:
* Create eventing with following CR and you will see MTBroker in the created pods.
```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeEventing
metadata:
  name: knative-eventing
  namespace: knative-eventing
spec:
  defaultBrokerClass: MTChannelBasedBroker
```
* With no `defaultBrokerClass` defined in the spec, `ChannelBasedBroker` is used in `config-br-defaults` configmap in `knative-eventing` namespace.

## Release Note

```release-note
NONE
```
